src/emulate.c

```diff
@@ -254,100 +254,90 @@ static inline bool insn_is_misaligned(uint32_t pc)
     );
 }
 
-/* execute a basic block */
-static bool emulate(riscv_t *rv, const block_t *block)
-{
-#if RV32_HAS(COMPUTED_GOTO)
-    static const void *dispatch_table[] = {
-#define _(inst) [rv_insn_##inst] = &&do_##inst,
-        RISCV_INSN_LIST
-#undef _
-    };
-
-#define DISPATCH()              \
-    /* enforce zero register */ \
-    rv->X[rv_reg_zero] = 0;     \
-    /* current IR */            \
-    ir = block->ir + index++;   \
-    /* jump */                  \
-    goto *dispatch_table[ir->opcode];
-
-/* clang-format off */
-#define _(inst, code)                    \
-    do_##inst: code                      \
-    /* step over instruction */          \
-    rv->PC += ir->insn_len;              \
-    /* increment the cycles CSR */       \
-    rv->csr_cycle++;                     \
-    /* all instructions have executed */ \
-    if (unlikely(index == n_insn))       \
-        return true;                     \
-    DISPATCH()
-/* clang-format on */
-#define EPILOGUE()
-
-#else /* !RV32_HAS(COMPUTED_GOTO) */
-#define DISPATCH()                          \
-    for (uint32_t i = 0; i < n_insn; i++) { \
-        ir = block->ir + i;                 \
-        /* enforce zero register */         \
-        rv->X[rv_reg_zero] = 0;             \
-        switch (ir->opcode) {
-/* clang-format off */
-#define _(inst, code)         \
-    case rv_insn_##inst: code \
-        break;
-#define EPILOGUE()                          \
-        }                                   \
-        /* step over instruction */         \
-        rv->PC += ir->insn_len;             \
-        /* increment the cycles csr */      \
-        rv->csr_cycle++;                    \
-    }                                       \
-    return true;
-/* clang-format on */
-#endif /* RV32_HAS(COMPUTED_GOTO) */
-
-    const uint32_t n_insn = block->n_insn;
-    rv_insn_t *ir;
-
-#if RV32_HAS(COMPUTED_GOTO)
-    /* current index in block */
-    uint32_t index = 0;
-#endif
+#define nextOP() \
+    ir++;        \
+    MUST_TAIL return ir->opfunc(rv, ir)
+#define RVOP(inst, code)                                              \
+    static bool rvop_##inst(riscv_t *rv UNUSED, rv_insn_t *ir UNUSED) \
```

Comment:
Replace `rvop_` with `do_` because we might extend IR for several micro-ops. Functions starting with `rvop_` would potentially confuse.

Reply:
Declare `rv_insn_t *ir` as `const`.

Reply:
We can modify `RISCV_INSN_LIST` in header `src/decode.h` and append branch information. i.e.,
```
/* RISC-V instruction list */
#define RISCV_INSN_LIST                    \
    _(nop)                                 \
    /* RV32I Base Instruction Set (opcode, can-branch) */       \
    _(lui, 0)                                 \
    _(auipc, 0)                               \
    _(jal, 1)                                 \
```

And then, we make use of `IIF` macro to specialize the cod sequence depending on its property (can-branch or not).

Function `insn_is_branch` can be simplified with X macro trick as well.

Reply:
Read [C Preprocessor tricks, tips, and idioms](https://github.com/pfultz2/Cloak/wiki/C-Preprocessor-tricks,-tips,-and-idioms) for pattern matching macros.